### PR TITLE
Summarizer correctness validation

### PR DESCRIPTION
**Pull request by tyler-cai-microsoft**
[AB#6490](https://dev.azure.com/fluidframework/internal/_workitems/edit/6490) and [AB#3400](https://dev.azure.com/fluidframework/internal/_workitems/edit/3400)

## Description
Doing the work for adding validation to check to see that the latest summary reference sequence number is the same when we summarize from the `RunningSummarizer` down to the `SummarizerNode` system, we ran into a bug in which the summary reference sequence numbers did not match. What was happening was for a single summarizer would submit a summary (S1), receive that summary's ack (S1Ack) in the `SummaryCollection`, but it wouldn't update the `SummarizerNode` system with that information as there was a summary lock. It would then start a new summary (S2) without processing (S1Ack) and thus we would be in an inconsitent state. 

For AB#6490, some logic was added to the `SummaryGenerator` which was already waiting for the summarize ack to actually process the summarize ack before considering its summary successfully completed.

## What's in the PR
This PR adds validation that the summary system from the `RunningSummarizer` down to the `SummarizerNode` system that the starting sequence numbers are correct. 

This PR also forces the `SummaryGenerator` to immediately update the `SummarizerNode` system with the summary ack it had already waited for.

### CODE REVIEW

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts [hunk 1]
```diff
@@ -14,6 +14,15 @@ export interface IRefreshSummaryResult {
 	isSummaryNewer: boolean;
 }
 
+export interface IStartSummaryResult {
+	/** The number of summarizerNodes at the start of the summary. */
+	nodes: number;
```
Comment (kian-thompson) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:19]: Is there value in tracking this number? May be useful for % of invalid nodes, but not too sure (since any number of invalid nodes is bad)

Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNodeUtils.ts:19]: It may be useful to see if it's a partial node tree issue or a whole node tree issue. I'm not sure what I'll see, but I'd like to know if part of the tree is updated, or none of the tree is updated.

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts [hunk 2]
```diff
@@ -243,11 +243,17 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
...
 
+		const immediatelyRefreshLatestSummaryAck =
+			this.mc.config.getBoolean("Fluid.Summarizer.immediatelyRefreshLatestSummaryAck") ??
+			false;
 		this.generator = new SummaryGenerator(
 			this.pendingAckTimer,
 			this.heuristicData,
 			this.submitSummaryCallback,
-			() => {
+			async (options: IRefreshSummaryAckOptions) => {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/runningSummarizer.ts:253]: Can you separate this logic into a separate PR. This will make it easier to review and also help back out or pick this part in case of any requirements.

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts [hunk 3]
```diff
@@ -243,11 +243,17 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
...
+				if (immediatelyRefreshLatestSummaryAck) {
+					await this.refreshLatestSummaryAckCallback(options);
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/runningSummarizer.ts:255]: This fix seems fine but it introduces another checkpoint into the system when we have other patterns to already handle these, i.e., `summarizingLock`.
The `summarizingLock` exists to ensure that summary submission and refreshing an ack are mutually exclusive and they happen in correct orders. I believe the real bug here is that in the `afterSummaryAction` function, there is an attempt to run the summaryHeuristics synchronously holding the same `summarizingLock` that was used to run the summary. That logic should be async to give any other process holding the lock to go through first.

Even with this fix, that bug exists where summary submission can hold the lock indefinitely.

Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/summary/runningSummarizer.ts:255]: I see what you mean. Ok that is really helpful. I'll look into this deeper.

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts (hunk 2 above)
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/summary/runningSummarizer.ts:253]: Will do

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summary/runningSummarizer.ts:255]: I would suggest trying that fix out. I am not 100% sure that will fix everything. Every time I look at the logic around summarizing + refreshing, I always learn something that I did not expect :-).

**Comment by tyler-cai-microsoft** [comment]
Migrated changes to this PR as this pr is old https://github.com/microsoft/FluidFramework/pull/19488
